### DISCUSSION TITLE
CSS-in-JS

### DISCUSSION THREAD

**Discussion by michoball** [wanted-9th-3team/frontend-study#47] [프론트 기술 질문 공간]
css-in-js 의 장 단점에 대해서 설명하고 사용하고 있는 라이브러리가 있다면 사용하는 이유를 알려주세요

**Comment by michoball** [comment]
css in js 는 js 코드 안에 css 스타일을 적용할 수 있도록 해주는 기술입니다. 
기존에 방식은 css 파일을 따로 만들고 html 안에 style 태그 같은 것으로 정의를 해야했습니다. 
React 의 경우 js 코드를 많이 분리해서 여러 파일에 다루기 때문에 하나 또는 몇 개의 css 파일에 모든 css 스타일을 정의할 경우  코드 가독성이 매우 떨어지고, 매번 새로운 class 나 id 명을 부여해야하는 단점이 있습니다. 

css in js의 경우 컴포넌트 별로 분리되어 스타일링을 할 수 있기 때문에 스타일과 관련된 class 나 id 값이 다른 컴포넌트와 겹처도 무방하고 또 가독성도 훨씬 좋습니다. 

저는 css in js로 styled components를 배워서 계속해서 사용하고 있는 중입니다. emotion 사용해봤는데 styled components 와 별로 다른 것 없어서 css in js를 프로젝트에서 사용한다면 어떤걸 사용해도 금방 익힐 것 같습니다.

**Comment by WooGie911** [comment]
CSS-in-JS는 단어 그대로 자바스크립트 코드에서 CSS를 작성하는 방식입니다.

CSS-in-JS 라이브러리는 대표적으로 Styled Components 와 Emotion이 있습니다. 

저는 Styled Components를 사용하는 이유는 자바스크립트 변수를 style에 사용할 수 있다는 점이  가장 큰 것 같습니다.
변수명을 style에 녹여냄으로서 더 쉽게 접근이 가능하고 가독성이 좋아져 사용중에 있습니다.
또한 .css파일을 CSS-in-CSS 로 관리할 경우 사용되지 않는 스타일을 쉽게 구분할 수 없기 때문에 CSS에 사용하지 않는 코드가 표시되는 경우가 많아집니다. 이러한 문제점도 해결 가능하다는 장점이 있습니다.

다만 번들 크기를 늘린다는 단점이 있지만 장점이 더 크다 생각하여 저는 CSS-in-JS에 긍정적으로 생각합니다.

**Comment by BaekYuri** [comment]
css in js는 css를 자바스크립트 코드 내에서 적용할 수 있도록 합니다.
저는 styled-component를 이용해 css in js를 경험해 본 적이 있는데, state가 변경될 때마다 css 속성을 바꾸는 방식으로 구현할 때 간편하여서 자주 사용하였습니다. 
css in js는 지역 스코프 스타일이기 때문에, 나중에 같은 클래스명을 다른 컴포넌트에서 똑같이 작성하더라도 독립된 속성으로 분리하여 의도치 않은 css 적용을 막을 수 있습니다. 또한 자바스크립트 내의 변수를 참조하여 사용할 수 있습니다.
단점으로는 번들 크기를 늘리고, 런타임 오버헤드를 더합니다. 컴포넌트가 렌더링 될 때 CSS-in-JS 라이브러리는 document에 삽입할 수 있는 일반 CSS로 스타일을 “직렬화”해야 합니다.
